### PR TITLE
Resolve `used-before-assignment` false negative when a function is defined under `TYPE_CHECKING` 

### LINKED ISSUE #10028: [used-before-assignment] Functions defined in type-checking guard not flagged when used later
In #9990, we [discovered](https://github.com/pylint-dev/pylint/pull/9990#issuecomment-2411463617) that extending the fix for false negatives for used-before-assignment for objects guarded behind type-checking blocks to include functions would cause a handful of false positives in the primer. So we backed out that change and opened this issue to investigate.

@zenlyj graciously offered to find time to look into it.

### PR DESCRIPTION
**Pull request by zenlyj**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |
|    | :sparkles: New feature |
|    | :hammer: Refactoring   |
|    | :scroll: Docs          |

## Description

Raise `used-before-assignment` when a function is defined under `TYPE_CHECKING` and used later.

Also extending the nonlocal handling to resolve the following false positive that was discovered by the [primer tests](https://github.com/pylint-dev/pylint/pull/9990#issuecomment-2411463617), where a variable usage references a nonlocal binding from an outer, enclosing frame.

These changes mainly include a search for nonlocal declarations in all enclosing frames of the variable usage. The idea is to rely on existing nonlocal checks and to only check outer frames if all nodes are filtered out by the consumer.

Refs #10028

### CODE REVIEW

**Comment by jacobtylerwalls** [comment]
Thanks so much @zenlyj, is this ready for review?

**Comment by zenlyj** [comment]
i don't think the current approach is correct, it silences the check in many cases and something like this is not flagged even though it would blow up in runtime:
```
def outer(callback):
    def inner():
        nonlocal callback
        def inner2():
            callback()
            def callback():
                pass
        inner2()
    inner()

def foo():
    print('foo')

outer(foo)
```

File: pylint/checkers/variables.py
Comment (zenlyj) [inline pylint/checkers/variables.py:683]: @jacobtylerwalls i was wondering if it would be better to have this nonlocal check in `VariablesChecker` instead of consumer. if it is in the checker, i suppose we need to check `consumer.consumed_uncertain` on top of `found_nodes` because of consumer filtering

**Comment by zenlyj** [comment]
@jacobtylerwalls this is ready for review, i tried out various approaches to handle the nonlocal false positive but settled for the current one, which reads better. let me know if there is anything i missed here, thank you!

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
LGTM, thank you, great work. Do you mind adding a fragment for the changelog ? (I'll also let Jacob review before merging).

**Comment by jacobtylerwalls** [comment]
I'll handle the Home Assistant false positive in another PR.

File: doc/whatsnew/fragments/10028.false_negative
Comment (jacobtylerwalls) [inline doc/whatsnew/fragments/10028.false_negative:3]:
```suggestion
Closes #10028
```

**Comment by zenlyj** [comment]
hey @jacobtylerwalls, i was just thinking about this case:
```
def nonlocal_in_outer_frame_ok(callback, condition_a, condition_b):
    def outer():
        nonlocal callback
        def inner():
            if condition_a:
                def inner2():
                    callback()  # possibly-used-before-assignment?
                inner2()
            else:
                if condition_b:
                    def callback():
                        pass
        inner()
    outer()
```
we should probably raise a message here?

File: doc/whatsnew/fragments/10028.false_negative
Comment (jacobtylerwalls) [reply doc/whatsnew/fragments/10028.false_negative:3]: Let's split that unused import comment into a new issue.

**Comment by jacobtylerwalls** [comment]
> we should probably raise a message here?

Good catch, I'm fine with opening a new issue for it. I think it should be real UBA, not "possibly".

**Comment by zenlyj** [comment]
it is worth noting that prior to my changes in this PR, we're emitting `possibly-used-before-assignment` for this snippet. but i feel like this will be tough to crack as it requires some form of control flow analysis?

**Review by jacobtylerwalls — APPROVED** [review]

File: pylint/checkers/variables.py
Comment (jacobtylerwalls) [inline pylint/checkers/variables.py:2242]: Hm. Is this the part of your change responsible for that new false negative? Or something else?

Comment (zenlyj) [reply pylint/checkers/variables.py:2242]: not quite, it is related to `has_nonlocal_binding()`. it does a search for nonlocals across all enclosing frames and silences the message when consumer filters out all nodes

Comment (jacobtylerwalls) [reply pylint/checkers/variables.py:2242]: Ah. So in the `while frame:` loop, can we stop the search after reaching an enclosing function? That would seem to solve the issue of the missing `nonlocal callback` in `inner()`.

Comment (zenlyj) [reply pylint/checkers/variables.py:2242]: that makes sense, what you're saying is we search up to the frame that encloses both the usage and assignment? in that case we need to look at both the "name" node and "found" node i guess

Comment (mbyrnepr2) [reply pylint/checkers/variables.py:2242]: Note similar logic in [this function](https://github.com/pylint-dev/pylint/blob/main/pylint/checkers/base/basic_error_checker.py#L393-L413) which could perhaps be refactored (into checkers/utils.py?) to serve both situations.

**Comment by Pierre-Sassoulas** [comment]
Hello @zenlyj. You've been doing a lot of great work for pylint lately, including this one, would you be interested in becoming a triager ? (I would normally have asked you privately but your commit mail is a no reply :) )

**Review by Pierre-Sassoulas — APPROVED** [review]

File: pylint/checkers/variables.py
Comment (zenlyj) [reply pylint/checkers/variables.py:2242]: thanks guys, let's hold off on merging this one. i'll make the updates later

**Comment by zenlyj** [comment]
> Hello @zenlyj. You've been doing a lot of great work for pylint lately, including this one, would you be interested in becoming a triager ? (I would normally have asked you privately but your commit mail is a no reply :) )

hello, i just dropped you an email, let's talk there

File: pylint/checkers/variables.py
Comment (zenlyj) [reply pylint/checkers/variables.py:2242]: resolving this thread, continuing with this at #10075

**Pull request merged**
